### PR TITLE
implement `takeover` for nydus fscache/fusedev mode

### PR DESCRIPTION
**Pull request by kevinXYin**
## Relevant Issue (if applicable)
https://github.com/dragonflyoss/nydus/issues/1421

## Details
This is based on https://github.com/dragonflyoss/nydus/pull/1434 , make the takeover feature can support both fusedev and fscache mode.

## Types of changes

_What types of changes does your PullRequest introduce? Put an `x` in all the boxes that apply:_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Go over all the following points, and put an `x` in all the boxes that apply._

- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -35,7 +35,7 @@ path = "src/lib.rs"
 anyhow = "1"
 clap = { version = "4.0.18", features = ["derive", "cargo"] }
 flexi_logger = { version = "0.25", features = ["compress"] }
-fuse-backend-rs = "^0.10.4"
+fuse-backend-rs = { git = "https://github.com/loheagn/fuse-backend-rs.git", branch = "vfs-persist" }
```
Comment (ccx1024cc) [inline Cargo.toml:38]: Use tag after published?

**Review by ccx1024cc — CHANGES_REQUESTED** [review]

**Review by ccx1024cc — DISMISSED** [review]
Smoke test is required.

File: Cargo.toml (hunk 1 above)
Comment (kevinXYin) [reply Cargo.toml]: thanks, fix

File: upgrade/src/backend/mod.rs [hunk 2]
```diff
@@ -0,0 +1,93 @@
+use std::{io, os::fd::RawFd};
```
Comment (jiangliu) [inline upgrade/src/backend/mod.rs:1]: Please add copyright header:)

File: upgrade/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+description = "Nydus Daemon Upgrade"
+authors = ["The Nydus Developers"]
+license = "Apache-2.0"
+homepage = "https://nydus.dev/"
+repository = "https://github.com/dragonflyoss/image-service"
+edition = "2021"
+
+[dependencies]
+sendfd = "0.4.3"
+snapshot = { git = "https://github.com/firecracker-microvm/firecracker", tag = "v1.4.1" }
```
[the remark this answers is not in the document]
Comment (kevinXYin) [reply upgrade/Cargo.toml]: Thanks, fixed

File: upgrade/src/backend/mod.rs (hunk 2 above)
Comment (kevinXYin) [reply upgrade/src/backend/mod.rs:5]: fixed

File: src/bin/nydusd/api_server_glue.rs [hunk 4]
```diff
@@ -279,6 +279,10 @@ impl ApiServer {
                         "{}",
                         e
                     ))))
+                } else if let Some(mut mgr_guard) = self.get_daemon_object()?.upgrade_mgr() {
```
Comment (jiangliu) [inline src/bin/nydusd/api_server_glue.rs:282]:
```
                } else {
                    if let Some(mut mgr_guard) = self.get_daemon_object()?.upgrade_mgr() {
                             mgr_guard.remove_blob_entry_state(&param.domain_id, &param.blob_id);
                    }
                    Ok(ApiResponsePayload::Empty)	                    Ok(ApiResponsePayload::Empty)
                }	                }
```

**Comment by jiangliu** [comment]
Otherwise LGTM, thanks for the great work:)

**Review by imeoer — APPROVED** [review]
Thanks for the work, let's add a smoke test case for validating fusedev failover later cc @loheagn .

**Review by hsiangkao — APPROVED** [review]

File: upgrade/src/backend/mod.rs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+/// StorageBackend trait is used to save and restore the fuse fds and fuse state data for online upgrade.
+pub trait StorageBackend: Send + Sync {
+    /// Save the fuse fds and fuse state data for online upgrade.
+    /// Returns the length of bytes of fuse state data.
+    fn save(&mut self, fds: &[RawFd], data: &[u8]) -> Result<usize>;
+
+    /// Restore the fuse fds and fuse state data for online upgrade.
+    /// Returns the length of bytes of fuse state data and the length of fds.
+    fn restore(&mut self, fds: &mut Vec<RawFd>, data: &mut Vec<u8>) -> Result<(usize, usize)>;
```
[the remark this answers is not in the document]
Comment (kevinXYin) [reply upgrade/src/backend/mod.rs]: Thanks, fixed

**Review by jiangliu — APPROVED** [review]

**Pull request merged**